cardano-config/src/Cardano/Config/TraceConfig.hs

```diff
@@ -0,0 +1,163 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Cardano.Config.TraceConfig
+  (
+    TraceConfig
+  , traceConfigEnabled
+  , traceConfigVerbosity
+  , traceConfigMute
+  , traceConfigParser
+
+  , TraceId
+  , traceEnabled
+
+  -- Per-trace toggles, alpha-sorted.
+  , traceAcceptPolicy
+  , traceBlockFetchClient
+  , traceBlockFetchDecisions
+  , traceBlockFetchProtocol
+  , traceBlockFetchProtocolSerialised
+  , traceBlockFetchServer
+  , traceChainDB
+  , traceChainSyncClient
+  , traceChainSyncBlockServer
+  , traceChainSyncHeaderServer
+  , traceChainSyncProtocol
+  , traceDnsResolver
+  , traceDnsSubscription
+  , traceErrorPolicy
+  , traceForge
+  , traceHandshake
+  , traceIpSubscription
+  , traceLocalChainSyncProtocol
+  , traceLocalErrorPolicy
+  , traceLocalHandshake
+  , traceLocalTxSubmissionProtocol
+  , traceLocalTxSubmissionServer
+  , traceLocalStateQueryProtocol
+  , traceMempool
+  , traceMux
+  , traceTxInbound
+  , traceTxOutbound
+  , traceTxSubmissionProtocol
+  )
+where
+
+import           Prelude (Show(..))
+import           Cardano.Prelude hiding (show)
+
+import           Data.Aeson
+import           Data.Aeson.Types (Parser)
+
+import           Cardano.BM.Data.Tracer (TracingVerbosity (..))
+
+import           Cardano.Config.Orphanage ()
+
+
+newtype TraceConfig = TraceConfig TraceConfig'
+
+instance Show TraceConfig where
+  show (TraceConfig tr) = show tr
+
+type TraceId = TraceConfig' -> Bool
+
+-- | Tracing configuration -- either disabled completely,
+--   or carries verbosity + fine-grained toggles.
+data TraceConfig'
+
+  -- | Intended to mirror Configuration with TurnOnLogging: False.
+  = TracingDisabled
+
+  -- | Detailed tracing options.  Except for verbosity, eaach option controls
+  --   a particular tracer.
+  | TraceOptions
+  -- Common part.
+  { traceVerbosity :: !TracingVerbosity
```

Comment:
That is a partial function. Production quality Haskell code should avoid using partial functions and more importantly should avoid adding them. Do we not still have a `-Wpartial-field` warnings enabled?

Reply:
@erikd, it's never exported, see the comment.  The partiality is strictly cointained to the module.

cc @dcoutts, as he approved this pattern previously.

Reply:
We should have ( and I thought we did have) `'-Wpartial-fields` enabled in all the cabal files. Why has this warning not triggerred? Are we not using `-Werror` ?

If we really are going to use this, then we first need to make sure `-Wpartial-fields` and `-Werror` are working as intended and than disable the partial fields error in this module using `-{-# OPTIONS_GHC -Wno-partial-field #-}`